src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will tie <see cref="Solution"/> and <see cref="RemoteHostClient.Session"/>'s lifetime together
+    /// so that one can handle those more easily
+    /// </summary>
+    internal sealed class SolutionAndSessionHolder : IDisposable
```

Comment:
Sorry, i keep looking at this and i just don't get it.  From the callers perspective they get something that is practically the same.  A disposable guy that they can "Invoke" on to perform communication with the OOP system.  When the caller is done they dispose this thing.

As a caller there is no difference to me between a "Session" and a "SolutionAndSessionHolder".  We already have the concept of "Session", i don't see why this sort of type needs to be leaked to the consumption side.